driver/driver.go

```diff
@@ -43,31 +48,40 @@ func (b bigQueryDriver) Open(uri string) (driver.Conn, error) {
 }
 
 func configFromUri(uri string) (*bigQueryConfig, error) {
-
 	if !strings.HasPrefix(uri, "bigquery://") {
 		return nil, fmt.Errorf("invalid prefix, expected bigquery:// got: %s", uri)
 	}
 
 	uri = strings.ToLower(uri)
-	path := strings.TrimPrefix(uri, "bigquery://")
-	fields := strings.Split(path, "/")
-
-	if len(fields) == 3 {
-		return &bigQueryConfig{
-			projectID: fields[0],
-			location:  fields[1],
-			dataSet:   fields[2],
-		}, nil
+	u, err := url.Parse(uri)
+	if err != nil {
+		return nil, fmt.Errorf("failed to parse connection string: %q", uri)
+	}
+
+	patternDatasetOnly := "/?*"
+	matchDatasetOnly, err := path.Match(patternDatasetOnly, u.Path)
+	if err != nil {
+		return nil, err
 	}
```

Comment:
What is the advantage of this vs comparing the count of parts after the split like it was before?

Reply:
Since I wanted to use a `*url.URL` mostly, the host becomes the first part of the string, then the path remains with either `location`, `dataset`, or both. By just adjusting it to handle 2 parts instead of 3 it won't behave the same as we'd had to add more validations. I thought it was easier to match by using these patterns, however, I'm checking the scenarios and I think one of them has different behavior so the current solcution is not consistent. 

What do you think about keeping the same logic as in the original function and using the `*url. URL` only to extract the query parameter?

```go
func configFromUriNew(uri string) (*bigQueryConfig, error) {
	if !strings.HasPrefix(uri, "bigquery://") {
		return nil, fmt.Errorf("invalid prefix, expected bigquery:// got: %s", uri)
	}

	u, err := url.Parse(uri)
	if err != nil {
		return nil, fmt.Errorf("failed to parse connection string: %q", uri)
	}

	scopes, err := getScopes(u.Query())
	if err != nil {
		return nil, err
	}

	path := strings.TrimPrefix(uri, "bigquery://")
	fields := strings.Split(path, "/")

	paramSuffix := "?" + u.RawQuery

	if len(fields) == 3 {
		return &bigQueryConfig{
			projectID: fields[0],
			location:  fields[1],
			dataSet:   strings.TrimSuffix(fields[2], paramSuffix),
			scopes:    scopes,
		}, nil
	}

	if len(fields) == 2 {
		return &bigQueryConfig{
			projectID: fields[0],
			location:  "",
			dataSet:   strings.TrimSuffix(fields[1], paramSuffix),
			scopes:    scopes,
		}, nil
	}

	return nil, fmt.Errorf("invalid connection string : %s", uri)
}


func getScopes(query url.Values) ([]string, error) {
	q := query.Get("scopes")
	q = strings.Trim(q, ",")
	scopes := []string{}
	if q != "" {
		scopes = strings.Split(q, ",")
	}

	return scopes, nil
}
```
Here I created a playground to check the cases: https://go.dev/play/p/R5wpSFXJnX8 please let me know if you catch something weird.

Reply:
Given that now we have an `url.URL` available we can use it for parsing most of the fields, initially I looked at the path patterns and I thought that they were kind of difficult to understand and that a count validation could make the code more readable. We can take advantage of the URL for cases like:

1. Validating the Scheme:

```go
	if u.Scheme != "bigquery" {
		return nil, fmt.Errorf("invalid prefix, expected bigquery:// got: %s", uri)
	}
```

2. Parsing the Path, this could be done with a default behavior of looking at the dataset at the last position and populating the location only when available:

```go
	fields := strings.Split(u.Path, "/")
	config := &bigQueryConfig{
		projectID: u.Hostname(),
		dataSet:   fields[len(fields)-1],
		scopes:    scopes,
	}

	if len(fields) == 2 {
		config.location = fields[0]
	}
```

The new helper for parsing the scopes looks great, I would switch the conditional to keep the happy path as left-indented as possible, this also reduces the allocation of an empty array that later will be overridden with another one:

```go
func getScopes(query url.Values) ([]string, error) {
	q := strings.Trim(query.Get("scopes"), ",")
	if q == "" {
		return []string{}, nil
	}
	return strings.Split(q, ","), nil
}
```

Reply:
I'm going to re-structure the code with your suggestions and see how it works 👀

Reply:
We did some changes and the function looks like this:
```go 
func configFromUriNew(uri string) (*bigQueryConfig, error) {
	u, err := url.Parse(uri)
	if err != nil {
		return nil, invalidConnectionStringError(uri)
	}

	if u.Scheme != "bigquery" {
		return nil, fmt.Errorf("invalid prefix, expected bigquery:// got: %s", uri)
	}

	if u.Path == "" {
		return nil, invalidConnectionStringError(uri)
	}

	fields := strings.Split(strings.TrimPrefix(u.Path, "/"), "/")
	if len(fields) > 2 {
		return nil, invalidConnectionStringError(uri)
	}

	config := &bigQueryConfig{
		projectID: u.Hostname(),
		dataSet:   fields[len(fields)-1],
		scopes:    getScopes(u.Query()),
	}

	if len(fields) == 2 {
		config.location = fields[0]
	}

	return config, nil
}

func getScopes(query url.Values) []string {
	q := strings.Trim(query.Get("scopes"), ",")
	if q == "" {
		return []string{}
	}
	return strings.Split(q, ",")
}

func invalidConnectionStringError(uri string) error {
	return fmt.Errorf("invalid connection string: %s", uri)
}
```

What do you think? @jags9415

Reply:
Looks great!